### PR TITLE
Insights operator up to date gathering

### PR DESCRIPTION
**Pull request by martinkunc**
This is suggesting how Insights Operator could become untied from release cycle and offer stability and dynamic changes when adding new data.

### CODE REVIEW

File: enhancements/insights/up-to-date-gathering.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+When CCX needs release faster: 
+- To identify affected customers (suggest Workaround), 
+- To find real feature usage, especially in in historical releases (because all that waiting on merge)
+- Fix IO bug
+
+
+### Goals
+
+As part of this enhancement we plan to do the following:
+* Release process change. On a fast-track release, publish Insights Operator fast-track changes during build to new quay tag for example 4.5-fast-track.
```
Comment (iNecas) [inline enhancements/insights/up-to-date-gathering.md:66]: Maybe worth splitting to two enhancements: one allowing the gathering part to run from different version of the image, and potentially other with changing the process to support one way of selecting the images.

File: enhancements/insights/up-to-date-gathering.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+only status. (This was discussed as an improvement to save memory in IO for larger datasets, like logs).
+It would have same security role as today.
+The proposal will be implemented in three phases, described in [Implementation phases](#implementation-phases).
+
+### Integrating with CVO
+
+There is still to be described how we are going to keep upgraded Operator Deployment after the image url change.
+Cluster Version Operator is reverting Operator Deployment config even if I just add a new container.
+There are two possible ways. Create separate, non CVO managed Deployment for Gatherer and make Gatherer separate Pod,
+or find some way how to make CVO keep changes in second container image after the Gatherer image update. This is still
```
Comment (iNecas) [inline enhancements/insights/up-to-date-gathering.md:103]: Even if CVO didn't revert the changes periodically, the insights-operator managing a separate "gathering" deployment and its images sounds more compelling to me than it updating it's own deployment.

File: enhancements/insights/up-to-date-gathering.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+*main* will be checking also health of *gatherer* container.
+Besides states it is reporting today, like when it cannot upload, it will be reporting
+error from *gatherer* call. When *host* detect an issue with *gatherer it will 
+update image back to its own image url, which will be stable release.
+*Main* would also need checking crashlooping state of *gatherer* for example if image cannot be downloaded.
+
+### Image selection
+
+There might be various ways how to select the good new image.
+The initial idea is to check for new versions check in quay.io. We would need to create new tag in quay,
```
Comment (iNecas) [inline enhancements/insights/up-to-date-gathering.md:118]: Other option I would consider is taking the latest image available in the particular stream.

File: enhancements/insights/up-to-date-gathering.md (hunk 2 above)
Comment (martinkunc) [reply enhancements/insights/up-to-date-gathering.md:103]: You are right. Because of CVO reconcillation, we would need to create separate Pod (I guess ReplicaSet) dynamically and then drive the Image version in it.

File: enhancements/insights/up-to-date-gathering.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+drive *gatherer*, but keep all the non-gathering logic, like uploading, status updates etc in it.
+It will be periodically checking for new versions in a 4.x-fast-track image on quay.io and eventually update the
+Insights Operator Gatherer container to latest 4.x-fast-track image url.
+It will not contain the Gathering methods and it will use Rest to reach *gatherer*, which will be 
+responsible for collection and saving data to storagePath (/var/lib/insights-operator) and this volume will be 
+shared among two containers.
+The *main* will instead of local Gather call remote Gather as Api, but will keep all the status handling and uploading
+in *main*.
+
+The Insights Operator *gatherer* will become just rest server on operators framework exposing Gather endpoint.
```
Comment (smarterclayton) [inline enhancements/insights/up-to-date-gathering.md:92]: I don't understand this.  The gatherer isn't a server, it's a client (to other things) and then writes things to disk.  The uploader takes things from disk and moves them to insights.

I do not see a justification for a server endpoint for the gatherer or for the uploader.

**Comment by bbrowning** [comment]
On the surface, if the Insights operator wants to ship updates separate from the OpenShift releases, the typical answer would be to ship Insights operator via OLM instead of bundled in OpenShift releases. I don't see it addressed in this proposal why that wasn't considered and why we're looking at a separate way to ship updated container images for this operator outside of either the typical CVO or OLM processes?

File: enhancements/insights/up-to-date-gathering.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+update image back to its own image url, which will be stable release.
+*Main* would also need checking crashlooping state of *gatherer* for example if image cannot be downloaded.
+
+### Image selection
+
+There might be various ways how to select the good new image.
+The initial idea is to check for new versions check in quay.io. We would need to create new tag in quay, 
+for example fast-track, and we would be checking
+```
+quay.io/openshift/origin-insights-operator:4-5-fast-track
```
Comment (jwforres) [inline enhancements/insights/up-to-date-gathering.md:121]: We need to support clusters where all images are mirrored to a private registry. I didn't see anything in the enhancement that talks about how it will work in those environments. All images in the core platform installation must be referenced by SHA today to work with ImageContentSourcePolicy.

File: enhancements/insights/up-to-date-gathering.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+```
+
+Note: I am using origin-insights-operator image, I am not sure if there is a better url to use for OCP image.
```
Comment (jwforres) [inline enhancements/insights/up-to-date-gathering.md:124]: All pull-spec translation for downstream (OCP) happens as part of our build processes. This is what the "image-references" file in the operator repos is for. But these processes are all designed to do translation to SHA pull-specs (no tags) for supporting mirrored registry environments.

File: enhancements/insights/up-to-date-gathering.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+- Fix IO bug
+
+
+### Goals
+
+As part of this enhancement we plan to do the following:
+* Release process change. On a fast-track release, publish Insights Operator fast-track changes during build to new quay tag for example 4.5-fast-track.
+The rest of release process will behave like today, release to 4.x.
+
+There will be two containers in Insights Operator Pod (*main* container with driver and *gatherer* container with Insights Operator).
```
Comment (smarterclayton) [inline enhancements/insights/up-to-date-gathering.md:69]: This sounds like trying to bypass being part of the payload (which updates at a certain rate) and seems to align with a lifecycle delivered via OLM.  This proposal should probably discuss "insights-operator delivered via OLM" as an alternative to this design, which would solve those problems more effectively.

File: enhancements/insights/up-to-date-gathering.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+### <a name="implementation-phases"></a> Implementation phases
+
+1. Implementation of *main* container and its orchestration in driver mode with execution of today’s Release version from Host
+with extracting the Gathering logic to *gatherer* and implementation of its host. At this phase driving will be disabled and 
+there will be no publication to fast-track tag.
+2. We will start publishing changes on fast-track release to Latest version available within the same stream 
+(4.4.z fast-track cluster might get gatherer from 4.4.z2 once available): 4.4.5 gatherer would still need 
+to go though the backporting process to verify it works with that particular stream. Reverting on failure to 
+"release - non fast-track" will be on place.
+3. Unification of fast-track streams. Instead of checking 4.5-fast-track on quay, driver will be looking for unified fast-track.
```
Comment (jwforres) [inline enhancements/insights/up-to-date-gathering.md:185]: This (and 2 above) sound like OLM channels. I'm concerned this enhancement is re-implementing OLM's out of band update model but for a single image within a component that is part of the core release payload. The underlying problem you are describing sounds like you need to install an OLM operator by default as part of a cluster installation.

File: enhancements/insights/up-to-date-gathering.md (hunk 6 above)
Comment (iNecas) [reply enhancements/insights/up-to-date-gathering.md:124]: What would be the possibilities? I guess extracting the image references from the latest ocp-release image would be one option (not very optimal one), having an API that would give the point to the right image would be the other. But I tent to agree that going an OLM way would cause less hassle.

File: enhancements/insights/up-to-date-gathering.md (hunk 7 above)
Comment (iNecas) [reply enhancements/insights/up-to-date-gathering.md:69]: Those are good points. Questions given we have an option for insights-operator to be delivered via OLM:

* would there be a way to have the OLM one installed by default?
* or would we keep the OLM one as optional, and still have the CVO present by default?
* or could the insights-operator run an image from OLM instead of the CVO one, but still the same pod?

Comment (iNecas) [reply enhancements/insights/up-to-date-gathering.md:69]: Alternatively, the insights-operator could collect payloads from external gatherers (with having it's own by default), and these gatherers being provided by OLM: it could be updated version of io itself, but there might be more, especially around other OLM operators for managed services.

File: enhancements/insights/up-to-date-gathering.md [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+The motivation is to be able to react faster on requirements for additional
+Insights Operator data requests. The benefit for customers would be faster
+identification of clusters based on newly discovered issues, identified by
+particular cluster configuration.
+The other potential slowdown might be caused by fact that administrators need
+to upgrade their clusters to deliver actual updates (either automatically or manually).
+
+Example of situation when the process is too slow:
+Two weeks ago we discovered an issue that for larger amount of generated alerts, IO Archives
+becomes too large (#148, CCXDEV-2673). We decided to fix it in all available versions (4.6, - 4.3)
```
Comment (russellb) [inline enhancements/insights/up-to-date-gathering.md:50]: What is 148?  I assume the CCXDEV reference is private, too.  Can you update this doc with publicly available references?

File: enhancements/insights/up-to-date-gathering.md (hunk 7 above)
Comment (martinkunc) [reply enhancements/insights/up-to-date-gathering.md:69]: I dont know much about lifecycle of OLM, if it would allow faster updates. But combined with failback to standard gathering it makes sense.

I was thinking if we would keep current Gathering + Saving archives to volume move to a shared volume, where whoever would be interested could provide more files to upload in some schema.
For example we have insights-2020-09-15-090843.tar.gz, but any other operator could leave here "some-other-data-[DATE].tar.gz", which could main IO just upload (latest by date).

File: enhancements/insights/up-to-date-gathering.md (hunk 8 above)
Comment (martinkunc) [reply enhancements/insights/up-to-date-gathering.md:185]: Good points jwforres, we would need to reconsider the approach to be more OLM specific. Thanks for all the good comments !

File: enhancements/insights/up-to-date-gathering.md (hunk 7 above)
Comment (martinkunc) [reply enhancements/insights/up-to-date-gathering.md:69]: I realised that the suggestion might look confusing when it is mixing deployment using standard CVO operated operator and also custom deployment with reaching quay manually. The reason for this was that CVO based operator would act "slow" branch where we would fail-back from "insights operator" downloaded by custom mechanism.

It looks like using CVO and OLM managed fast track would be more suitable. Would you see any problems when CVO classic operator would install operator using OLM and let it gather (eventually fail-back to CVO, released one) ?

**Comment by markmc** [comment]
@sdodson, @bparees, and I met @iNecas and others responsible for insights-operator to briefly discuss this problem space last week.

They explained how - in the months since this enhancement was first proprosed - it's become ever more clear that in order to achieve their support experience objectives, they need to be able to react to newly discovered issues in real-world clusters by implementing new Insights rules, and these rules oftern require additional data to be gathered from clusters.

In other words, the motivation for this enhancement remains relevant - if the only avenue for rolling out data gathering changes across releases is via an OCP release image update, then we'll be severly restricted in our ability to respond quickly.

At a high level, separating the the data collection code from the rest of the operator makes the most sense to me. If we can minimize the risk of a data gathering change from disrupting the cluster, then allowing the operator to dynamically pull down those changes seems reasonable.

/assign sdodson bparees markmc

**Comment by markmc** [comment]
Some specific points to consider when refreshing the proposal:

- Make this enhancement very specifically about updating the data gathering code in order to allow Insights rules to identify a newly disocvered issue on real-world clusters

- Detail the prior art of how updates are pushed for data gathering code for the RHEL insights-client

- Explain how you will minimize the risk of a data gathering change causing disruption to the cluster

- Take a close look at how must-gather works and whether you can build on it - e.g. the way a gather image gets run and the data copied out; the way multiple gather images are shipped; how it might be ueful if users could use `oc adm must-gather` to run the insights gather image to take a look at what would be uploaded

- Consider whether 'oc adm inspect' is something you build on (like must-gather does) - perhaps it could be enhanced to gain anonymization support, for example

- Discuss where how new data gathering code will be published - e.g. should the operator hit the insights API to find the latest gather image it should use

- If you want to keep fixes for the operator itself in scope for the enhancement, then OLM seems like the right way to go. If you decide not to go the OLM route, add an explanation to the alternatives section about it.

**Comment by sdodson** [comment]
>     * Discuss where how new data gathering code will be published - e.g. should the operator hit the insights API to find the latest gather image it should use

And if we lose the strong assurances provided by a signed release manifest by fetching code from a remote endpoint how will we provide the same code signing guarantees?

I think it'd be helpful to elaborate on why there are situations where quick turn around on gathering updates is the only option. If we had infinite understanding of the product and ability to deliver rules in the next minor release are there scenarios where we still cannot predict data we may need to gather in the future?

Are we just substituting flexibility in gathering code for better system design? For instance, thinking about recent sudo vulnerabilities if that were something Insights attempted to identify, designing a system that maps RHCOS versions to the version of sudo present in those versions is better than writing a rule which looks at sudo package directly because it provides a platform for all package based vulnerability alerting without fleet updates.

If we pursue either OLM or fetching gathering rules from a remote location, what facilities do we have to halt that rollout in the event that we discover problems with the new rules? CVO managed rollouts can be halted via Cincinnati and in the future will have more advanced progressive rollout characteristics.

Some of the questions from reviewers focused on ability to run in disconnected environments, however it seems that Insights provides no autonomous reporting, do we know whether or not it's common for clusters to require disconnected image mirroring but have access to our backend services?

**Comment by tremes** [comment]
@markmc , @sdodson Thanks for the update. I will reopen this topic in my team (and with Ivan) and I think we will create a new enhancement proposal because Martin Kunc has left the company.
